File: src/RNetPi.Core/RNet/ZoneSourcePacket.cs
using System;

namespace RNetPi.Core.RNet;

/// <summary>
/// Represents an RNet Zone Source packet
/// </summary>
public class ZoneSourcePacket : DataPacket
{
    public ZoneSourcePacket()
    {
        MessageType = 0x00;
    }

    public byte GetControllerID()
    {
        return SourceControllerID;
    }

    public byte GetZoneID()
    {
        return SourcePath.Length > 2 ? SourcePath[2] : (byte)0;
    }

    public byte GetSourceID()
    {
        return Data.Length > 0 ? Data[0] : (byte)0;
    }

    /// <summary>
    /// Creates a ZoneSourcePacket from a DataPacket
    /// </summary>
    public static ZoneSourcePacket FromPacket(DataPacket dataPacket)
    {
        if (dataPacket.MessageType != 0x00)
        {
            throw new ArgumentException("Cannot create ZoneSourcePacket from packet with MessageType != 0x00");
        }

        var zoneSourcePacket = new ZoneSourcePacket();
        dataPacket.CopyToPacket(zoneSourcePacket);
        return zoneSourcePacket;
    }
}

File: src_js/rnet/packets/ZoneSourcePacket.js
const DataPacket = require("./DataPacket");

class ZoneSourcePacket extends DataPacket {
    requiresHandshake() {
        return true;
    }

    getControllerID() {
        return this.sourceControllerID;
    }

    getZoneID() {
        return this.sourcePath[2];
    }

    getSourceID() {
        return this.data.readUInt8(0);
    }
}

ZoneSourcePacket.fromPacket = function(dataPacket) {
    if (dataPacket instanceof DataPacket) {
        const zoneSourcePacket = new ZoneSourcePacket();
        dataPacket.copyToPacket(zoneSourcePacket);
        return zoneSourcePacket;
    }
    else {
        throw new TypeError("Cannot create ZoneSourcePacket with anything other than a DataPacket");
    }
}

module.exports = ZoneSourcePacket;
